### PR TITLE
[recipe, doc] feat: add full-weight Qwen-Image OCR FlowGRPO recipe

### PR DESCRIPTION
**Pull request by AndyZhou952**
### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

- Add a full-weight Qwen-Image OCR FlowGRPO training script.
- Document OOM tuning guidance for rollout TP, reward TP, and micro-batch sizes.
- Make rollout/reward TP configurable in the existing Qwen-Image OCR LoRA scripts.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `vllm_omni`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `diffusion`, `omni`, `tests`, `docker`
  - If this PR involves multiple modules, separate them with `,` like `[diffusion, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][diffusion, fsdp] feat: new rollout scheduler`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

Experiments are performed on 4 NVIDIA H200s w/ no cfg (i.e., `true_cfg_config=1.0`, set optimizer buffer to fp32, image resolution 384x384, lr=1e-5, clip range=1e-5). Verifying version w/ cfg required more cards to avoid OOM.

Validation reward (reached 0.95 at step 270; reached 0.90 at step 210):
image

Time per step (~230s):
image

Critic-related metrics:
- The reward increased steadily from 0.1 to 0.9. The std mean is at peak when the mean reward is around 0.5.
image

Actor-related metrics:
The (total) clip range increased from ~10% to ~30%, as recommended.
image

validation  

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update the documentation.
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl-omni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

File: examples/flowgrpo_trainer/run_qwen_image_ocr.sh
Comment (SamitHuang) [inline examples/flowgrpo_trainer/run_qwen_image_ocr.sh:1]: Pls attach the training results (convergence curve & throughput etc, and wandb link if possible) to the Performance Reference doc

File: docs/algo/flowgrpo.md [hunk 1]
```diff
@@ -252,6 +252,14 @@ For reward models that are expensive to evaluate (e.g., a VLM judge), the reward
 bash examples/flowgrpo_trainer/run_qwen_image_ocr_lora_async_reward.sh
 ```
 
+### Full Model Training
+
+We have provided a script to enable full-weight Qwen-Image OCR training. The example is runnable on 4 NVIDIA H200 GPUs with `true_cfg_scale=1.0`; enabling CFG with `true_cfg_scale > 1.0` requires more GPUs.
```
Comment (SamitHuang) [inline docs/algo/flowgrpo.md]: update with non-cfg?

**Review by SamitHuang — CHANGES_REQUESTED** [review]

File: examples/flowgrpo_trainer/run_qwen_image_ocr.sh
Comment (AndyZhou952) [reply examples/flowgrpo_trainer/run_qwen_image_ocr.sh:1]: Included in the performance.md. Can consider updating it once we have some more robust results (i.e. w/ lr 3e-5, verification on full resolution 512 etc. to be consistent with the LoRA setting)

File: docs/algo/flowgrpo.md (hunk 1 above)
Comment (AndyZhou952) [reply docs/algo/flowgrpo.md]: done

File: docs/algo/performance.md
Comment (SamitHuang) [inline docs/algo/performance.md:1]: pls also attach clip frac ratio curve, thx

File: examples/flowgrpo_trainer/run_qwen_image_ocr.sh [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+
+python3 -m verl_omni.trainer.diffusion.main_flowgrpo \
+    algorithm.adv_estimator=flow_grpo \
+    data.train_files=$ocr_train_path \
+    data.val_files=$ocr_test_path \
+    data.train_batch_size=32 \
+    data.max_prompt_length=256 \
+    actor_rollout_ref.model.path=$model_name \
+    actor_rollout_ref.actor.optim.lr=3e-5 \
```
Comment (zhtmike) [inline examples/flowgrpo_trainer/run_qwen_image_ocr.sh]: 1e-5 or 3e-5?

Comment (AndyZhou952) [reply examples/flowgrpo_trainer/run_qwen_image_ocr.sh]: thanks! it is 1e-5 indeed. Updated

File: docs/algo/performance.md
Comment (AndyZhou952) [reply docs/algo/performance.md:1]: updated

File: verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py [hunk 3]
```diff
@@ -126,6 +126,18 @@ async def run_headless(self, args: argparse.Namespace):
     def _get_wake_up_tags(self) -> list[str]:
         return ["weights"]
 
+    async def _sleep_hybrid(self):
+        """Preserve non-actor pipeline weights during hybrid training sleep.
+
+        vLLM-Omni diffusion pipelines include components such as the text
```
Comment (zhtmike) [inline verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py:132]: this should be more generic I think. In future, if the trainer side incorperate the whole components of the model, then it is not a true.

Leave a TODO here

Comment (AndyZhou952) [reply verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py:132]: Good points. Added

File: docs/algo/performance.md [hunk 4]
```diff
@@ -30,10 +30,34 @@ Below are reference benchmark results for VeRL-Omni training runs.
...
 </div>
 
 - `qwen_image_ocr_lora`: sync reward, 4 GPUs (`run_qwen_image_ocr_lora.sh`)
 - `qwen_image_ocr_lora_async_reward`: async reward on a dedicated 5th GPU (`run_qwen_image_ocr_lora_async_reward.sh`)
 
 > **Note:** Reward curves may differ from the references above mainly due to rollout-side stochasticity: diffusion rollouts sample random latents/noise, and the example scripts do not fix the data seed, so prompt ordering can vary between runs.
+
+## FlowGRPO: non-CFG Full Model Training on Qwen-Image OCR
+
+> Experiments used NVIDIA H200 GPUs, `ppo_micro_batch_size_per_gpu` 8, lr 1e-5, clip_ratio 1e-5, image resolution 384x384, optimizer buffer fp32. The other parameters are consistent with the LoRA setting.
```
Comment (SamitHuang) [inline docs/algo/performance.md]:
```suggestion
> Experiments used NVIDIA H200 GPUs, `ppo_micro_batch_size_per_gpu` 8, lr 1e-5, clip_ratio 1e-5, image resolution 384x384, optimizer state fp32. The other parameters are consistent with the LoRA setting.
```

File: verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py [hunk 5]
```diff
@@ -126,6 +126,19 @@ async def run_headless(self, args: argparse.Namespace):
...
+        encoder and VAE that are loaded by the rollout server, but are not part
+        of the trainable actor and therefore are not included in full-model
+        weight syncs. Use level-1 sleep so those weights are offloaded and can
+        be restored on wake-up instead of discarded by level-2 sleep.
+        """
+        # TODO: make it more generic in the future when the trainer side incorperate the whole components of the model.
```
Comment (zhtmike) [inline verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py]:
```suggestion
        # TODO (Andy):  consider to change it back to sleep level 2 once the trainer side incorporates the whole components of the model.
```

**Review by zhtmike — APPROVED** [review]

File: verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py [hunk 6]
```diff
@@ -126,6 +126,20 @@ async def run_headless(self, args: argparse.Namespace):
...
+        """Preserve non-actor pipeline weights during hybrid training sleep.
+
+        vLLM-Omni diffusion pipelines include components such as the text
+        encoder and VAE that are loaded by the rollout server, but are not part
+        of the trainable actor and therefore are not included in full-model
+        weight syncs. Use level-1 sleep so those weights are offloaded and can
+        be restored on wake-up instead of discarded by level-2 sleep.
+        """
+        # TODO: make it more generic in the future when the
+        #  trainer side incorporates the whole components of the model.
```
Comment (zhtmike) [inline verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py]:
```suggestion
        # TODO (andy): use `sleep_level=2` in the future when the
        #  trainer side incorporates the whole components of the model.
```

File: examples/flowgrpo_trainer/README.md [hunk 7]
```diff
@@ -92,6 +92,18 @@ For reward models that are expensive to evaluate (e.g., a VLM judge), the reward
 bash examples/flowgrpo_trainer/run_qwen_image_ocr_lora_async_reward.sh
 ```
 
+Ulysses sequence parallelism shards the sequence dimension across GPUs to reduce per-GPU memory. A ready-to-use 4-GPU SP=2 LoRA example is provided:
+
+```bash
+bash examples/flowgrpo_trainer/run_qwen_image_ocr_lora_sp2.sh
```
Comment (SamitHuang) [inline examples/flowgrpo_trainer/README.md:98]: why put lora + sp2 in this PR? Is it verified?

Comment (AndyZhou952) [reply examples/flowgrpo_trainer/README.md:98]: for consistency - previous PR updated the docs but did not update this README.md page. Should have an explanation here. Thus added along here with the full model changes

**Review by SamitHuang — APPROVED** [review]
LGTM
